Added support for ITE IT51XXX SoC and related drivers

GTLin08 opened the pull request:
This PR adds support for the ITE IT51XXX SoC along with its associated drivers, enabling full functionality for various peripherals.

Included drivers:
Core drivers: Interrupt, Clock, Timer, GPIO, Pinctrl, Flash

Additionally, this PR includes board support for it51526_evb, which has been tested for boot and driver functionality.

GTLin08 committed:
soc/ite/ec: common: Modify the format to comply with check_compliance.py

Modify the format to comply with check_compliance.py.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
soc: ITE: Add ITE it51xxx SoC

Add support for ITE it51xxx SoC.
NOTE: it51526aw is not support RISCV_ISA_EXT_C.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
board: ite/it515xx: Add board of it515xx

Add support for ITE it515xx_evb board.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers/interrupt: Add interrupt and wake-up control drivers of it51xxx

Add interrupt and wake-up control drivers for ITE it51xxx series.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers/serial: Add ITE UART wrapper to enable serial driver of ns16550

Add ITE UART wrapper to enable serial driver of ns16550.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers/timer: Add timer driver of it51xxx

Add timer driver for ITE it51xxx series.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers/clock: Add clock drivers of it51xxx

Add clock drivers for ITE it51xxx series.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers/flash: Enable flash controller for it51xxx series

Enable flash controller for ITE it51xxx series.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers/pinctrl: Enable pinctrl driver for it51xxx series

Enable pinctrl driver for ITE it51xxx series.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers/gpio: Add GPIO driver of it51xxx

Add GPIO driver for ITE it51xxx series.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
soc: Separate it8xxx2 related files from the common folder

Some files should be separated from the common folder to
each soc folder for future expansion of the chip.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

henrikbrixandersen reviewed (dismissed):
Please split this into a base PR (this PR) adding the following, minimal support:

> Included drivers: Core drivers: Interrupt, Clock, Timer, GPIO, Pinctrl
> 
> Additionally, this PR includes board support for it51526_evb, which has been tested for boot and driver functionality.

These can be submitted as individual PRs after the initial, basic support has been added:

> Peripheral support: Flash, Serial(UART), I2C, I3C, PWM, Tachometer, WDT, ADC, Voltage comparator, Input driver

Please see https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#defining-smaller-prs

GTLin08 committed:
soc: ITE: ilm: Enable instruction memory for it51xxx series

Enable instruction memory for ITE it51xxx series.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 commented on dts/riscv/ite/it51xxx.dtsi:
```diff
@@ -0,0 +1,69 @@
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		cpu0: cpu@0 {
```
Done.

GTLin08 commented on dts/riscv/ite/it51xxx.dtsi:
```diff
@@ -0,0 +1,69 @@
+			compatible = "ite,it51xxx-gctrl";
+			reg = <0x00f02000 0x100>;
+		};
+
```
Done.

GTLin08 commented on soc/ite/ec/it51xxx/Kconfig:
```diff
@@ -0,0 +1,42 @@
+	select FLASH
+	imply XIP
+
+config SOC_IT51526AW
```
Remove.

GTLin08 commented on soc/ite/ec/it51xxx/Kconfig:
```diff
@@ -0,0 +1,42 @@
+	hex
+	default $(dt_chosen_reg_size_hex,$(DT_CHOSEN_ZEPHYR_FLASH))
+	help
+	  Total size of writable flash.
```
This is not used, remove.

GTLin08 commented on soc/ite/ec/it51xxx/Kconfig.defconfig.it51526aw:
```diff
@@ -0,0 +1,6 @@
+
+if SOC_IT51526AW
+
+endif
```
Drop.

GTLin08 commented on soc/ite/ec/it51xxx/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,39 @@
+if SOC_SERIES_IT51XXX
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 32768
```
Fixed.

GTLin08 commented on soc/ite/ec/it51xxx/linker.ld:
```diff
@@ -0,0 +1,469 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
```
This linker file is not just a copy of the Zephyr linker script, modifications were made to support the IT51XXX SoC. The key changes include:

- Exception Section Alignment (256 bytes):
 The _EXCEPTION_SECTION_NAME section is aligned to 256 bytes as required by IT51XXX.

- Instruction Local Memory (ILM) Handling:
__ilm_ram_start and __ilm_ram_end define the mapped ILM region in RAM, which ensures the ILM code is copied from Flash to RAM at runtime.
The ILM mapping must be 4K-aligned (ALIGN(0x1000)) to meet IT51XXX memory constraints.

GTLin08 commented on soc/ite/ec/it51xxx/soc.c:
```diff
@@ -0,0 +1,119 @@
+
+	return 0;
+}
+SYS_INIT(ite_it51xxx_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
Fixed.

GTLin08 commented on soc/ite/ec/it51xxx/Kconfig:
```diff
@@ -39,4 +40,18 @@ config SOC_IT51XXX_FLASH_SIZE_BYTES
+	  predictable than executing from Flash directly, and some code (such as code
+	  that writes to the internal Flash) must execute out of RAM.
+
+config ILM_MAX_SIZE
+	int "ILM Size in kB"
+	default 4
```
Since the ILM feature is compatible with the ILM of IT8XXX2, modifying it would affect the related IT8XXX2 driver. I believe this change should be handled in a separate PR after this one is merged. Therefore, in this PR, we will temporarily retain the ILM_MAX_SIZE config.

The PR for the planned modification is as follows:
https://github.com/GTLin08/zephyr/commit/6fe50490b21d547be191b185886d58582d4a4ffd

GTLin08 commented on boards/ite/it515xx_evb/doc/WinFlashTool_P1.jpg:
These photos have been deleted. There is the same description in [it8xxx2_evb](https://docs.zephyrproject.org/latest/boards/ite/it82xx2_evb/doc/index.html#flashing), and a link has been added.

GTLin08 commented on boards/ite/it515xx_evb/it515xx_evb.dts:
```diff
@@ -0,0 +1,79 @@
+
+	leds {
+		compatible = "gpio-leds";
+		led0: led_0 {
```
Done.

GTLin08 commented on boards/ite/it515xx_evb/it515xx_evb.dts:
```diff
@@ -0,0 +1,79 @@
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		slot0_partition: partition@0 {
```
Done.

GTLin08 commented on boards/ite/it515xx_evb/it515xx_evb.dts:
```diff
@@ -0,0 +1,79 @@
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
```
The IT51XXX runs programs using internal flash and does not use partitions. However, Zephyr tests under the path test/drivers/flash/common/ require the storage_partition node.

GTLin08 commented on boards/ite/it515xx_evb/it515xx_evb_defconfig:
```diff
@@ -0,0 +1,23 @@
+CONFIG_GEN_IRQ_VECTOR_TABLE=n
+
+# General Kernel Options
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=32768
```
Done.

GTLin08 commented on boards/ite/it515xx_evb/it515xx_evb_defconfig:
```diff
@@ -0,0 +1,23 @@
+# General Kernel Options
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=32768
+CONFIG_XIP=y
+CONFIG_INIT_STACKS=y
```
Remove this config as it increases initialization time and memory usage.

GTLin08 commented on boards/ite/it515xx_evb/it515xx_evb_defconfig:
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2025 ITE Corporation. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_GEN_IRQ_VECTOR_TABLE=n
```
This config disabled by default in the riscv architecture, so it can be removed.

GTLin08 commented:
> Please split this into a base PR (this PR) adding the following, minimal support:

I have split the changes as requested. This PR now contains the minimal support.

fabiobaltieri commented on soc/ite/ec/it51xxx/soc.c:
```diff
@@ -0,0 +1,116 @@
+
+bool cpu_idle_not_allowed(void)
+{
+	return !!(atomic_get(&cpu_idle_disabled));
```
I think that the `!!` is redundant, seince the return type is a bool the compiler should do this for you

GTLin08 replied:
Fixed.

fabiobaltieri commented on soc/ite/ec/it51xxx/chip_chipregs.h:
```diff
@@ -298,4 +299,7 @@ struct gctrl_it51xxx_regs {
 #define IT51XXX_GCTRL_ALTIE       BIT(4)
 
+/* Alias gpio_it8xxx2_regs to gpio_it51xxx_regs for compatibility */
+#define gpio_it8xxx2_regs gpio_it51xxx_regs
```
this is a bit of a weird way of doing it, would it make sense to fix up the original driver to use a generic name?

GTLin08 replied:
I have already updated the original related structures and registers to use a generic name, as seen in the commit: https://github.com/zephyrproject-rtos/zephyr/pull/87095/commits/77a54558980a8b4e5d081d7b85d29d85dccd6a28 .

fabiobaltieri commented on drivers/clock_control/clock_control_it51xxx.c:
```diff
@@ -0,0 +1,204 @@
+{
+	const struct clock_control_it51xxx_config *const config = dev->config;
+	struct ite_clk_cfg *clk_cfg = (struct ite_clk_cfg *)(sub_system);
+	uint8_t tmp_mask = 0;
+
+	/* CGCTRL3R, bit 6, must always write a 1. */
+	tmp_mask |= (clk_cfg->ctrl == IT51XXX_ECPM_CGCTRL3R_OFF) ? 0x40 : 0x00;
```
could you not just set it directly?

GTLin08 replied:
Fixed.

GTLin08 committed:
drivers: ITE: Use generic name instead of specific chip name

Use generic name for structure in driver instead of specific chip name
for better compatibility.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 commented:
Hi @henrikbrixandersen, @nordicjm,
I’ve made the changes requested and would appreciate your review. Let me know if you have any further comments. Thanks!

GTLin08 commented:
Hi @fabiobaltieri ,
Thanks for your review! While working on the other peripheral, we noticed that `clock_control_it51xxx_on`/`off `was writing to the wrong register. The write should be to `config->ecpm_base + clk_cfg->ctrl` instead of `config->ecpm_base`. I have updated the clock:[commit ](https://github.com/zephyrproject-rtos/zephyr/pull/87095/commits/abc8249cc2195c80ec2b56fdea41a100d1874f77)accordingly.

GTLin08 commented:
Rebase the PR and to fix conflicts in chip_chipreg.h.

Dino-Li commented:
@fabiobaltieri please revisit.
@nordicjm any comments would be appreciated.

nordicjm commented on dts/riscv/ite/it51xxx-pinctrl-map.dtsi:
```diff
@@ -0,0 +1,324 @@
+	adc0_ch0_gpi0_default: adc0_ch0_gpi0_default {
+		pinmuxs = <&pinctrli 0 IT8XXX2_ALT_FUNC_1>;
+	};
+	adc0_ch1_gpi1_default: adc0_ch1_gpi1_default {
```
newlines between child nodes

GTLin08 replied:
Done.

nordicjm commented on boards/ite/it515xx_evb/doc/it515xx_evb_and_debug_card.jpg:
doesn't look done

GTLin08 replied:
Converted to webp.

fabiobaltieri commented:
needs a rebase

GTLin08 commented:
> doesn't look done

@nordicjm, Done, related photos have been converted from jpg to webp files.

GTLin08 commented:
> needs a rebase

Rebase the PR and to fix conflicts in clock_control/CMakeLists.txt

Dino-Li commented:
@fabiobaltieri @nordicjm please revisit.

nordicjm reviewed (approved):
Build system changes OK
